panels/user-accounts/um-app-permissions.c

```diff
@@ -0,0 +1,787 @@
+  gtk_image_set_pixel_size (GTK_IMAGE (w), size);
+  gtk_container_add (GTK_CONTAINER (box), w);
+  g_object_unref (icon);
+
+  /* App name label */
+  w = g_object_new (GTK_TYPE_LABEL,
+                    "label", app_name,
+                    "hexpand", TRUE,
+                    "xalign", 0.0,
+                    NULL);
+  gtk_container_add (GTK_CONTAINER (box), w);
+
+  /* Switch */
+  w = gtk_switch_new ();
+  gtk_container_add (GTK_CONTAINER (box), w);
+
+  gtk_widget_show_all (box);
+
+  /* Fetch status from AccountService */
+  allowed = epc_app_filter_is_appinfo_allowed (self->filter, app);
+
+  gtk_switch_set_active (GTK_SWITCH (w), allowed);
+  g_object_set_data_full (G_OBJECT (w), "GAppInfo", g_object_ref (app), g_object_unref);
+
+  blacklisted = g_hash_table_contains (self->blacklisted_apps, app);
+  if (allowed && blacklisted)
+    g_hash_table_remove (self->blacklisted_apps, app);
+  else if (!allowed && !blacklisted)
+    g_hash_table_add (self->blacklisted_apps, g_object_ref (app));
+
+  g_signal_connect (w, "notify::active", G_CALLBACK (on_switch_active_changed_cb), self);
+
+  return box;
+}
+
+static gint
+compare_app_info_cb (gconstpointer a,
+                     gconstpointer b,
+                     gpointer      user_data)
+{
+  GAppInfo *app_a = (GAppInfo*) a;
+  GAppInfo *app_b = (GAppInfo*) b;
+
+  return g_utf8_collate (g_app_info_get_display_name (app_a),
+                         g_app_info_get_display_name (app_b));
+}
+
+static void
+on_set_age_action_activated (GSimpleAction *action,
+                             GVariant      *param,
+                             gpointer       user_data)
+{
+  GsContentRatingSystem rating_system;
+  UmAppPermissions *self;
+  const gchar * const * entries;
+  const guint *ages;
+  guint age;
+  guint i;
+
+  self = UM_APP_PERMISSIONS (user_data);
+  age = g_variant_get_uint32 (param);
+
+  rating_system = get_content_rating_system (self->user);
+  entries = gs_utils_content_rating_get_values (rating_system);
+  ages = gs_utils_content_rating_get_ages (rating_system);
+
+  /* Update the button */
+  for (i = 0; entries[i] != NULL; i++)
+    {
+      if (ages[i] == age)
+        {
+          gtk_button_set_label (self->restriction_button, entries[i]);
+          break;
+        }
+    }
+
+  g_assert (entries[i] != NULL);
+
+  g_debug ("Selected age: %u", age);
+
+  self->selected_age = age;
+
+  schedule_update_blacklisted_apps (self);
+}
+
+/* GObject overrides */
+
+static void
+um_app_permissions_finalize (GObject *object)
+{
+  UmAppPermissions *self = (UmAppPermissions *)object;
+
+  g_cancellable_cancel (self->cancellable);
+  g_clear_object (&self->apps);
+  g_clear_object (&self->cancellable);
+  g_clear_object (&self->system_installation);
+  g_clear_object (&self->user);
+  g_clear_object (&self->user_installation);
+
+  g_clear_pointer (&self->blacklisted_apps, g_hash_table_destroy);
+  g_clear_pointer (&self->filter, epc_app_filter_unref);
```

Comment:
What about `action_group`? It’s marked as `owned` in the `struct` at the top of the file.

And you might want to clear `blacklist_apps_source_id`, or assert that it is zero.

Reply:
Ah, the `blacklist_apps_source_id` is cleared in `dispose()`. Very good. Maybe whack a `g_assert (self->blacklist_apps_source_id == 0)` in here to make it really clear.